### PR TITLE
feat(core): add TDF encrypt/decrypt commands and authentication to platform SDK

### PR DESCRIPTION
**Pull request by jakedoublev**
Closes https://github.com/opentdf/otdfctl/issues/105
Closes https://github.com/opentdf/otdfctl/issues/75

Will handle https://github.com/opentdf/otdfctl/issues/129 separately

### CODE REVIEW

**Comment by b-long** [comment]
Hey @jakedoublev , cool feature!  I've been playing with this over the last few days and noticed that it just started working (at least with the configuration I'm using).   

On my machine, I'm standing up the platform according to the platform "Run" section, here: https://github.com/opentdf/platform#run .

Once the platform is running, I've been doing the following:

```bash
# Checkout this branch
git checkout feat/tdf-demo

# Build `otdfctl`
make

# Get Help
cp otdfctl.yaml target/ && pushd target && { ./--darwin-arm64 --help ; } && popd

# Authenticate
cp otdfctl.yaml target/ && pushd target && { ./--darwin-arm64 auth client-credentials --client-id opentdf --client-secret secret --log-level DEBUG ; } && popd

# Encrypt
#
# NOTE: This will create a file 'sensitive.txt.tdf'
cp otdfctl.yaml target/ && pushd target && { echo "Some string" | ./--darwin-arm64 encrypt || echo "Failed to encrypt" ; } && popd

# Decrypt
cp otdfctl.yaml target/ && pushd target && { ./--darwin-arm64 decrypt sensitive.txt.tdf -o decrypted.txt || echo "Failed to decrypt" ; } && popd
```

Previously I'd get an error about DPoP:

```
"time":"2024-04-23T11:58:54.970023-04:00","level":"DEBUG","msg":"LoadConfig: file and key not provided, using default file","config file":""}
{"time":"2024-04-23T11:58:55.216295-04:00","level":"DEBUG","msg":"getting new access token"}
{"time":"2024-04-23T11:58:55.216339-04:00","level":"DEBUG","msg":"Building DPoP Proof"}
 ERROR    Failed to decrypt file: reader.doPayloadKeyUnwrap failed:  splitKey.rewrap failed:error making request to kas: error making rewrap request: rpc error: code = Unauthenticated desc = unauthenticated
```

As of this morning, I'm not getting this error message! 🎉 

I'm guessing one of the [recent platform commits](https://github.com/opentdf/platform/commits/main/) (e.g. maybe [this DPoP change](https://github.com/opentdf/platform/pull/617)?) has unlocked this for me.  Of course, that's a total guess, and doesn't really matter 😃 

That said, there does seem to be a small bug in the `decrypt` command.  I am specifying `-o decrypted.txt`, but it looks like the output is coming to STDOUT.  Have you seen that behavior?

**Comment by jrschumacher** [comment]
@b-long Yes the DPoP change makes it optional since we've been having some issues supporting DPoP across IdPs given that it is [still a proposed standard](https://datatracker.ietf.org/doc/html/rfc9449).

Thanks for trying this out and giving feedback!

**Comment by b-long** [comment]
@jakedoublev @jrschumacher Here comes a "stupid question" from me, so apologies in advance as I'm just learning golang.

Is this a reasonable fix to `-o`?  https://github.com/opentdf/otdfctl/pull/128

If not, can you provide guidance about how it can be improved?

**Comment by jakedoublev** [comment]
@b-long Thanks for your engagement, and glad you're excited about this feature. :) I think your proposed fix for `-o` should be sufficient, but I actually wonder if we may want to skip this flag altogether and make it a convention to write stdout to a file instead (i.e. `./otdfctl decrypt sensitive.txt.tdf > output.txt`).

Flag `-o` pros and cons:
🟩 : decrypted content and decrypted content alone are written to the output file
🟨 : if writing stdout to a file, the only logs should be debug, and running in debug mode will have known side effects on stdout
🟥 : if encrypt reads from stdin, it seems like decrypt should write to stdout
🟥 : reserves the `-o` flag namespace permanently

Would the flow `./otdfctl decrypt sensitive.txt.tdf > output.txt` meet your needs, in which case I could remove the flag altogether and update the manual/docs?

**Comment by b-long** [comment]
Personally, I like having `-o` as an option.  Especially , given this pro:

> 🟩 : decrypted content and decrypted content alone are written to the output file

That said, I also think it's nice to have an option to write to STDOUT.  I like both 🙂 

RE: Your other points:

> 🟨 : if writing stdout to a file, the only logs should be debug, and running in debug mode will have known side effects on stdout

I'm not too worried about the logging behavior.  Except, if you remove the `-o` feature.  If `-o` (the feature) is removed then I  worry about the risk of log content being written and my decrypted file malformed (e.g. anything that isn't just text; `.png`, `.pdf`, `.mov` files) 

> 🟥 : if encrypt reads from stdin, it seems like decrypt should write to stdout

I agree, piping is a useful feature 👍 

> 🟥 : reserves the `-o` flag namespace permanently

Happy to use any flag, but I think the feature (write to a file, explicitly) is useful. 👍

**Review by b-long — DISMISSED** [review]

**Review by b-long — APPROVED** [review]

File: docs/man/encrypt/_index.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      description: A filename and extension that will be TDFd (i.e. '-o password.txt' -> 'password.txt.tdf', default 'sensitive.txt.tdf' or <file>.tdf) and placed in the current working directory.
+      default: '' # default is set dynamically to allow filename parsing
+    - name: attr
+      shorthand: a
+      description: Attribute value Fully Qualified Names (FQNs, i.e. 'https://example.com/attr/attr1/value/value1') to apply to the encrypted data.
+---
+
+Build a Trusted Data Format (TDF) with encrypted content from a specified file or input from stdin utilizing OpenTDF platform.
+
+Examples:
```
Comment (jrschumacher) [inline docs/man/encrypt/_index.md:17]:
```suggestion
## Examples
```

File: docs/man/decrypt/_index.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+      shorthand: o
+      description: 'The file destination for decrypted content to be written.'
+      default: ''
+---
+
+Decrypt a Trusted Data Format (TDF) file and output the contents to stdout or a file in the current working directory.
+
+The first argument is the TDF file with path from the current working directory being decrypted (default 'sensitive.txt.tdf').
+
+Examples:
```
Comment (jrschumacher) [inline docs/man/decrypt/_index.md:16]:
```suggestion
## Examples
```

File: docs/man/decrypt/_index.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+```bash
+# default to sensitive.txt.tdf, then print to stdout
+otdfctl decrypt
+
+# specify the TDF to decrypt then output decrypted contents
+otdfctl decrypt hello.txt.tdf # print to stdout
+otdfctl decrypt hello.txt.tdf > hello.txt # consume stdout to write to hello.txt file
+otdfctl decrypt hello.txt.tdf -o hello.txt # write to hello.txt file
```
Comment (jrschumacher) [inline docs/man/decrypt/_index.md:25]: Can we add a piping example
```suggestion
otdfctl decrypt hello.txt.tdf -o hello.txt # write to hello.txt file

# piping support
cat hello.txt.tdf | otdfctl decrypt > hello.txt
```

File: docs/man/encrypt/_index.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+---
+title: Encrypt file or stdin as a TDF
+command:
+  name: encrypt [file]
+  flags:
+    - name: out
+      shorthand: o
+      description: A filename and extension that will be TDFd (i.e. '-o password.txt' -> 'password.txt.tdf', default 'sensitive.txt.tdf' or <file>.tdf) and placed in the current working directory.
```
Comment (jrschumacher) [inline docs/man/encrypt/_index.md:8]: I don't think this should be the default behavior. I would expect the default behavior is output to stdout. So we can support something like this

```
tar cv /secrets | gzip | otdfctl encrypt | s3 cp - s3://bucket/secrets.tar.gz.tdf
```

Comment (jakedoublev) [reply docs/man/encrypt/_index.md]: Great feedback. Updated in newer commits.

**Review by jrschumacher — APPROVED** [review]

**Pull request merged**